### PR TITLE
Improve error recovery for null coalescing assignment operator

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Part of compiler part of https://github.com/dotnet/roslyn/issues/73894 (the issue also mentiones operators like `+=`, `-=` etc., but I focused only on `??=` in this PR to keep it relatively small)

While working on PR I found several inconsistencies in what compiler APIs return in normal and error scenarios. These are marked with `// INCONSISTENCY:` comments in tests. We need to decide what the desired behavior should be and either:
- Remove/alter the comments if in the given case the behavior is decided to be expected
- Make an issue (or several ones) for cases, that can/should be improved ...
- ... or alternatively fix them in this PR (which I don't like given that the total size of this PR is already around ~1000 lines with actual changes taking just ~20 lines of the codebase)

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
Cna I have a review here please?

**Comment by DoctorKrolic** [comment]
Can I get a review here please?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 1]
```diff
@@ -4170,17 +4169,31 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
...
                 return new BoundNullCoalescingAssignmentOperator(node, leftOperand, rightOperand, CreateErrorType(), hasErrors: true);
             }
+            // If right operand is bad, make sure conversion is in place for better error recovery
+            else if (rightOperand.HasAnyErrors)
+            {
+                var rightOperandTargetType = leftType switch
+                {
+                    { IsReferenceType: true } or { IsValueType: false, TypeKind: TypeKind.TypeParameter } => leftType,
+                    { IsValueType: true } when leftType.IsNullableType() => leftType.GetNullableUnderlyingType(),
+                    _ => CreateErrorType()
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: Not sure we should be doing this conversion when there's an error type here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs [hunk 2]
```diff
@@ -186,5 +187,925 @@ void M(int? a)
...
+
+            var typeInfo = model.GetTypeInfo(node);
+            Assert.Equal("T", typeInfo.Type.ToTestDisplayString());
+            Assert.Equal("T", typeInfo.ConvertedType.ToTestDisplayString());
+
+            var symbolInfo = model.GetSymbolInfo(node);
+            Assert.Null(symbolInfo.Symbol);
+            // INCONSISTENCY: Having type itself as a symbol info candidate for a constructor invocation node is odd.
+            // If there were no errors symbol info would be completely empty (no symbol & no candidates).
+            // Given that we know type of the expression, shouldn't symbol info be identical for normal and error cases?
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:355]: I don't really understand this comment. What do you mean that if there were no errors there'd be no symbol info? Is there an example somewhere here to better illustate the inconsistency?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs [hunk 3]
```diff
@@ -186,5 +187,925 @@ void M(int? a)
...
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+
+            var descendantNodes = tree.GetCompilationUnitRoot().DescendantNodes();
+            var node = descendantNodes.OfType<ConditionalExpressionSyntax>().Single();
+
+            var typeInfo = model.GetTypeInfo(node);
+            // INCONSISTENCY: if conditional expression didn't have any errors, it would have a `null` type.
+            // Shouldn't this be preserved given that an expression is target-typed and have no natural type anyway?
+            // Moreover, target-typed tuples and collection expressions already behave that way
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:682]: Error cases do error things. Having them be fully consistent with non-error scenarios is nice, but ultimately not a requirement.

**Review by 333fred — COMMENTED** [review]
The general approach is fine to me, but I don't really understand one of the inconsistencies pointed out. Also, I suspect that if we didn't generate a conversion in the error type scenario, a few things might naturally work out to be more consistent.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: Some expressions (e.g. unconverted object expression) need conversion no matter what. I remember when I implemented this I tried to avoid this error type to error type conversion, but that triggered an assert later, because sometimes unconverted expressions where not handled correctly, e.g. `notNullableValueType ??= new()`. So I left it there

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: Sure, but those should just go through our standard mechanisms for this, `BindToNaturalTypeForErrorRecovery` or whatever it's called.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs (hunk 2 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:355]: > What do you mean that if there were no errors there'd be no symbol info?

If there were no errors on the right side of the expression returned symbol info would be just a `default`:
devenv_eJ2aWsH9u5

> Is there an example somewhere here to better illustate the inconsistency?

`new T()` is syntactically a constructor. Therefore I would expect `GetSymbolInfo` return a constructor metho symbol or nothing when it cannot be determined. Here `GetSymbolInfo` returns nothing for a happy path, but for error recovery starts to (unexpectedly to me) return _type symbol_ T. This is inconsistent to what `GetSymbolInfo` usually returns for syntactic constructor calls, isn't it?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:355]: Is there a test case to illustrate the whole behavior here?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: Tried locally. This broke 2 tests: `ErrorRecovery_CollectionExpression_NotNullableValueType` and `ErrorRecovery_TargetTypedInterpolatedString_NotNullableValueType`. In the first one _converted_ type changed from error type to `null`. In the second one _convrted_ type changed form error to `string`. While for the first case I am not sure which output is better, for the second one I prefer current PR output (error). What do you think?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: I probably prefer the change.

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: > I probably prefer the change.

Sorry, which one? Original PR change (error) or your suggested change (string)? The latter should also be what a compiler from `main` currently outputs

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4188]: The latter.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs (hunk 2 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:355]: Looked for `new T()` text in the codebase and it seems like there aren't any tests which specificly verify public semantic model API output for it. Gonna add a few

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:393]: Added equivalent cases without error to demonstrate, that symbol info in such scenarios is always empty

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs [hunk 4]
```diff
@@ -314,6 +314,44 @@ void M()
...
+
+            var descendantNodes = tree.GetCompilationUnitRoot().DescendantNodes();
+            var node = descendantNodes.OfType<ImplicitObjectCreationExpressionSyntax>().Single();
+
+            var typeInfo = model.GetTypeInfo(node);
+            Assert.Equal("T", typeInfo.Type.ToTestDisplayString());
+            Assert.Equal("T", typeInfo.ConvertedType.ToTestDisplayString());
+
+            var symbolInfo = model.GetSymbolInfo(node);
+            Assert.True(symbolInfo.IsEmpty);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: This frankly seems like a bug. @CyrusNajmabadi, is this the behavior you'd expect from this test?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: i agree.  that seems odd.  what's the reason for that?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: if we had `new T()` instead, do we get the same result?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: Oh, I see what's going on here. This is the generic `where T : new()` case. There literally isn't a constructor symbol here.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: @DoctorKrolic, I believe that explains what you were seeing: yes, it is expected that `where T : new()` won't show a symbol here, because there isn't one _to_ show.

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: > it is expected that `where T : new()` won't show a symbol here, because there isn't one _to_ show.

Interesting. Given that `new()` implies, that type has a parameterless constructor, I thought compiler makes a constructor symbol for such type parameters.

With that question resolved, we don't expect anything to be returned in an equivalent error case then, right?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: > we don't expect anything to be returned in an equivalent error case then, right?

Correct, nothing should be returned. That being said, if there are some particularly odd cases that do, imo it's fine. Error scenarios are best effort.

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: > There literally isn't a constructor symbol here

Out of curiosity: why is that?

> Given that `new()` implies, that type has a parameterless constructor, I thought compiler makes a constructor symbol for such type parameters.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:347]: > I thought compiler makes a constructor symbol for such type parameters.

It doesn't. What metadata symbol would that reflect?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 5]
```diff
@@ -4170,17 +4169,31 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
...
+            if (leftOperand.HasAnyErrors || (rightOperandTargetType is null && rightOperand.HasAnyErrors))
             {
                 leftOperand = BindToTypeForErrorRecovery(leftOperand);
                 rightOperand = BindToTypeForErrorRecovery(rightOperand);
                 return new BoundNullCoalescingAssignmentOperator(node, leftOperand, rightOperand, CreateErrorType(), hasErrors: true);
             }
+            // If right operand is bad, but we know its type, make sure conversion is in place for better error recovery
+            else if (rightOperand.HasAnyErrors)
+            {
+                var conversion = GenerateConversionForAssignment(rightOperandTargetType, rightOperand, diagnostics, ConversionForAssignmentFlags.CompoundAssignment);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:4191]: We still need to ensure that the left side is bound for error recovery.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 6]
```diff
@@ -3389,6 +3389,12 @@ public ILocalSymbol GetDeclaredSymbol(CatchDeclarationSyntax catchDeclaration)
                         {
                             if (resultKind == LookupResultKind.NotCreatable)
                             {
+                                if (expr.Type is TypeParameterSymbol)
+                                {
+                                    resultKind = LookupResultKind.Empty;
+                                    return symbols;
+                                }
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:3396]: Feels like we should separate this out into a different PR. It's not related to the `??=` change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs [hunk 7]
```diff
@@ -186,5 +187,1060 @@ void M(int? a)
...
+
+            var descendantNodes = tree.GetCompilationUnitRoot().DescendantNodes();
+            var node = descendantNodes.OfType<ImplicitObjectCreationExpressionSyntax>().Single();
+
+            var typeInfo = model.GetTypeInfo(node);
+            Assert.True(typeInfo.Type.IsErrorType());
+            Assert.True(typeInfo.ConvertedType.IsErrorType());
+
+            var symbolInfo = model.GetSymbolInfo(node);
+            Assert.Null(symbolInfo.Symbol);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:308]: Do we understand why this differs from the nullable version? Is it because the whole `??=` is bogus?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:308]: Not saying that must change, just trying to understand the why.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs (hunk 3 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:832]: Sooo, do we just left everything as is and I should just delete all these comments?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs (hunk 7 above)
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:308]: If type of LHS is a reference type, we know that according to semantics of `??=` operator RHS must be of the same type. If type of LHS is a nullable value type, we know, that type of RHS must be an underlying type of that nullable. However, if LHS is a non-nullable value type we have no idea what type is RHS. In this particular case it is syntactically a constructor, but what constructor? Of `Nullable<T>` or of `T` or what? So given that we cannot answer this question, we just return empty symbol info.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullCoalesceAssignmentTests.cs:831]: For the cases that make sense to do so, yes. Let's get the `new T()` change merged, then come back and update all of these baselines and see where we're at.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 6 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: @DoctorKrolic did you open a separate PR for this? Would prefer to merge that first.

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Please take a look: https://github.com/dotnet/roslyn/pull/74675
